hollow/src/main/java/com/netflix/hollow/core/index/HollowPrefixIndex.java

```diff
@@ -0,0 +1,438 @@
+package com.netflix.hollow.core.index;
+
+import com.netflix.hollow.core.memory.encoding.FixedLengthElementArray;
+import com.netflix.hollow.core.memory.pool.ArraySegmentRecycler;
+import com.netflix.hollow.core.memory.pool.RecyclingRecycler;
+import com.netflix.hollow.core.read.engine.HollowReadStateEngine;
+import com.netflix.hollow.core.read.engine.HollowTypeStateListener;
+import com.netflix.hollow.core.read.engine.object.HollowObjectTypeReadState;
+import com.netflix.hollow.core.schema.HollowObjectSchema;
+import com.netflix.hollow.core.schema.HollowSchema;
+
+import java.util.BitSet;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * This class builds a prefix index. A prefix index can be used to build applications like auto-complete, spell checker.
+ */
+public class HollowPrefixIndex implements HollowTypeStateListener {
+
+    private HollowReadStateEngine readStateEngine;
+    private String type;
+    private String fieldPath;
+
+    private Tst prefixIndex;
+    private volatile Tst prefixIndexVolatile;
+    private RecyclingRecycler memoryRecycle;
```

Comment:
This should reference an `ArraySegmentRecycler` -- it could be either a `RecyclingRecycler` (which is a memory pool), or a `WastefulRecycler` (which simply creates new array segments and drops them on the floor when they are returned).